src/query/src/region_query.rs

```diff
@@ -15,14 +15,33 @@
 use std::sync::Arc;
 
 use async_trait::async_trait;
+use common_meta::node_manager::NodeManagerRef;
 use common_query::request::QueryRequest;
 use common_recordbatch::SendableRecordBatchStream;
+use partition::manager::PartitionRuleManagerRef;
+use session::ReadPreference;
 
 use crate::error::Result;
 
+/// A factory to create a [`RegionQueryHandler`].
+pub trait RegionQueryHandlerFactory: Send + Sync {
+    /// Build a [`RegionQueryHandler`] with the given partition manager and node manager.
+    fn build(
+        &self,
+        partition_manager: PartitionRuleManagerRef,
+        node_manager: NodeManagerRef,
+    ) -> RegionQueryHandlerRef;
+}
+
+pub type RegionQueryHandlerFactoryRef = Arc<dyn RegionQueryHandlerFactory>;
+
 #[async_trait]
 pub trait RegionQueryHandler: Send + Sync {
-    async fn do_get(&self, request: QueryRequest) -> Result<SendableRecordBatchStream>;
+    async fn do_get(
+        &self,
+        read_preference: ReadPreference,
+        request: QueryRequest,
+    ) -> Result<SendableRecordBatchStream>;
```

Comment:
Should we add a read_preference field to the query request?

Reply:
The region server will also reuse the `QueryRequest`.